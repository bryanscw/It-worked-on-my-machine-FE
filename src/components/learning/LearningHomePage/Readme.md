### Example:

```js
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { MemoryRouter as Router } from 'react-router-dom';
import 'bootstrap/dist/css/bootstrap.css';
import 'bootstrap/dist/js/bootstrap.js';

import { METHODS } from '../../../redux/ducks/apiHelper.js';
import { withReduxRouter } from '../../../utils/mock.js';

const initialState = {
    authReducer: {
        user: {
            "name": "Bob"
        }
    },
    topicsReducer: {
        items: [
            {
                "id": 1,
                "title": "Addition",
                "description": "Putting 2 and 2 together",
            },
            {
                "id": 2,
                "title": "Subtraction",
                "description": "Removing numbers from each other",
            }
        ],
        isLoading: {
            [METHODS.LIST]: false
        },
        hasFailed: {
            [METHODS.LIST]: false
        }
    }
}

withReduxRouter(<LearningHomePage />, initialState)
```