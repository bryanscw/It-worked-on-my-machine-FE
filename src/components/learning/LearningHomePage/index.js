import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import Loader from '../../common/Loader';
import { listTopics, selectTopics, selectTopicsLoading, selectTopicsFailed } from '../../../redux/ducks/topics';
/**
 * This component displays the learning home page for a student. It contains a list of study topics to choose from
 */
class LearningHomePage extends Component {
    componentDidMount() {
        this.props.listTopics();
    }

    render() {
        const {
            topicsLoading,
            topicsFailed,
            topics,
            user
        } = this.props;

        if (topicsLoading)
            return <Loader />
        
        return (
            <div className="container">
                <h1>Welcome {user.name}!</h1>
                <h2>Topics</h2>
                {
                    topics.length !== 0 && !topicsFailed
                        ? topics.map((topic) => (
                            <div href="#" className="card mb-4" key={topic.id}>
                                <div className="card-body">
                                    <Link to={`/topics/${topic.id}`}>
                                        <h3 className="card-title">{topic.title}</h3>
                                    </Link>
                                    <p className="card-text">{topic.description}</p>
                                </div>
                            </div>
                        ))
                        : <p>No topics found.</p>
                }
            </div>
        );
    }
}

LearningHomePage.propTypes = {
    /** The currently logged in user */
    user: PropTypes.object.isRequired,

    /** A boolean to determine if the topics are still being loaded by the `listTopics` action creator (true: still loading, false: fully loaded) */
    topicsLoading: PropTypes.bool.isRequired,
    /** A boolean to determine if the topics failed to be loaded by the `listTopics` action creator (true: still loading or failed to load, false: successful load) */
    topicsFailed: PropTypes.bool.isRequired,
    /** An array of topic objects loaded by the `listTopics` action creaor */
    topics: PropTypes.array.isRequired,
    /** An action creator for listing topics */
    listTopics: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
    user: state.authReducer.user,
    topicsLoading: selectTopicsLoading(state),
    topicsFailed: selectTopicsFailed(state),
    topics: selectTopics(state),
});

const dispatchers = {
    listTopics,
};

export default connect(mapStateToProps, dispatchers)(LearningHomePage);
