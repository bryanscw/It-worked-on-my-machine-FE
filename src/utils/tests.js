import React from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';

import rootReducer from '../redux/rootReducer';

export const renderWithRedux = (component, initialState = {}) => {
    const store = createStore(
        rootReducer,
        initialState,
        applyMiddleware(thunk),
    )
    return <Provider store={store}>{component}</Provider>;
}

export const renderWithRouter = (component, route = '/') => {
    const history = createMemoryHistory({ initialEntries: [route] });
    return <Router history={history}>{component}</Router>
}

export const renderWithReduxRouter = (component, initialState = {}, route = '/') => {
    return renderWithRedux(renderWithRouter(component, route), initialState);
}