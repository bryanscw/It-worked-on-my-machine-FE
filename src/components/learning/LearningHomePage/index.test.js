import React from 'react'
import { cleanup, waitForElement } from '@testing-library/react'
import axiosMock from 'axios'

import { renderWithReduxRouter } from '../../../utils/tests.js';
import LearningHomePage from './index.js';

jest.mock('axios');

beforeEach(() => {
    axiosMock.get.mockResolvedValue({
        data: {
            content: [
                {
                    id: 1,
                    title: "Addition",
                    description: "Putting 2 and 2 together."
                },
                {
                    id: 2,
                    title: "Subtraction",
                    description: "Removing one from another."
                }
            ]
        }
    });
});

afterEach(() => {
    cleanup();
    jest.clearAllMocks();
});

it('should take a snapshot', async () => {
    const { asFragment, getByText } = renderWithReduxRouter(<LearningHomePage />);
    await waitForElement(() => getByText('Topics'));
    expect(asFragment()).toMatchSnapshot();
})

it('should load topics', () => {
    const { getByTestId } = renderWithReduxRouter(<LearningHomePage />); 
    expect(getByTestId('loader')).toBeTruthy();
});

it('should load and display the topics', async () => {
    const { getByText } = renderWithReduxRouter(<LearningHomePage />, {
        authReducer: {
            user: {
                name: 'Bob'
            }
        }
    });

    const welcome = await waitForElement(() => getByText('Welcome Bob!'));
    expect(welcome).toBeTruthy();

    const topics = await waitForElement(() => getByText('Topics'));
    expect(axiosMock.get).toHaveBeenCalledTimes(1);
    expect(topics).toBeTruthy();
});

it('should not display edit or delete buttons', async () => {
    const { getByText, container } = renderWithReduxRouter(<LearningHomePage />);
    
    await waitForElement(() => getByText('Topics'));
    
    const editButton = container.querySelector('[data-icon="edit"]');
    expect(editButton).toBeFalsy();

    const deleteButton = container.querySelector('[data-icon="trash"]');
    expect(deleteButton).toBeFalsy();
});