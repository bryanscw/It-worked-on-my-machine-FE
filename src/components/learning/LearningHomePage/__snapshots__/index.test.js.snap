// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h1>
      Welcome !
    </h1>
    <h2>
      Topics
    </h2>
    <div
      class="card mb-4"
      href="#"
    >
      <div
        class="card-body"
      >
        <a
          href="/topics/1"
        >
          <h3
            class="card-title"
          >
            Addition
          </h3>
        </a>
        <p
          class="card-text"
        >
          Putting 2 and 2 together.
        </p>
      </div>
    </div>
    <div
      class="card mb-4"
      href="#"
    >
      <div
        class="card-body"
      >
        <a
          href="/topics/2"
        >
          <h3
            class="card-title"
          >
            Subtraction
          </h3>
        </a>
        <p
          class="card-text"
        >
          Removing one from another.
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
